src/cascadia/TerminalConnection/AzureConnection.cpp

```diff
@@ -974,17 +980,54 @@ namespace winrt::Microsoft::Terminal::TerminalConnection::implementation
         auto uri{ fmt::format(L"{}terminals?cols={}&rows={}&version=2019-01-01&shell={}", _cloudShellUri, _initialCols, _initialRows, shellType) };
 
         WWH::HttpStringContent content{
-            L"",
+            L"{}",
             WSS::UnicodeEncoding::Utf8,
             // LOAD-BEARING. the API returns "'content-type' should be 'application/json' or 'multipart/form-data'"
             L"application/json"
         };
 
-        const auto terminalResponse = _SendRequestReturningJson(uri, content);
+        const auto terminalResponse = _SendRequestReturningJson(uri, content, WWH::HttpMethod::Post(), Windows::Foundation::Uri(_cloudShellUri));
         _terminalID = terminalResponse.GetNamedString(L"id");
 
+        auto socketUri = terminalResponse.GetNamedString(L"socketUri");
+
+        // we have to do some post-handling to get the proper socket endpoint
+        // the logic here is based on the way the cloud shell team itself does it
+        winrt::hstring finalSocketUri;
+        std::wstring wCloudShellUri{ _cloudShellUri };
+
+        if (wCloudShellUri.find(L"servicebus") == std::wstring::npos)
+        {
+            // wCloudShellUri does not contain the word "servicebus", we can just use it to make the final URI
+
+            // remove the "https://" from the cloud shell URI
+            std::wstring https{ L"https://" };
+            wCloudShellUri.replace(wCloudShellUri.find(https), https.length(), L"");
+
+            finalSocketUri = fmt::format(FMT_COMPILE(L"wss://{}terminals/{}"), wCloudShellUri, _terminalID);
+        }
+        else
+        {
+            // if wCloudShellUri contains the word "servicebus", that means the returned socketUri is of the form
+            // wss://ccon-prod-westus-aci-03.servicebus.windows.net/cc-AAAA-AAAAAAAA//aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
+            // we need to change it to:
+            // wss://ccon-prod-westus-aci-03.servicebus.windows.net/$hc/cc-AAAA-AAAAAAAA/terminals/aaaaaaaaaaaaaaaaaaaaaa
+
+            // get the substring up until the ".net"
+            const std::wstring_view wSocketUri{ socketUri };
+            const auto dotNetPos = wSocketUri.find(L".net") + 4;

```

Comment:
I'm still a little worried about the case when this returns `npos` (which is the max. value of a `size_t`), because then `+4` will cause it to underflow to `dotNetPos == 3` and not report any error. You could consider using a `std::wstring` here and `.append()` to build your own URI step-by-step without relying on a single `fmt::format` call. That way you could wrap these into `if` conditions.

But that's more of a minor concern for me.